lib/runner.js

```diff
@@ -142,52 +159,72 @@ Runner.prototype.onExit = function onExit(code, signal) {
 
   debug('Exit of %s with status %s stopped? %s', this, status, this.stopped);
 
+  this.child.exited = true;
+  this.child.exitStatus = code || signal;
+
   this.child = null;
   this.emit('exit', status);
   this._restart(status);
 };
 
 // Callback with exit status if killed, with nothing if already dead.
-Runner.prototype.kill = function kill(callback) {
-  if (!callback) {
-    callback = function() {};
-  }
+Runner.prototype.kill = function kill(signame, callback) {
+  var child = this.child;
 
-  if (!this.child) {
+  if (!child) {
+    // No child associated.
     process.nextTick(callback);
-    return this;
-  }
 
-  var signame = 'SIGTERM';
+  } else if (child.exited) {
+    // Child has already had it's exit event.
+    process.nextTick(function() {
+      callback(child.exitStatus);
+    });
 
-  debug('Kill process %s with %s', this.child.pid, signame);
+  } else {
+    // The child hasn't emitted an 'exit' event yet.
+    // Associate the callback - if specified - with the exit callback.
+    if (callback) {
+      child.once('exit', function(code, signal) {
+        callback(code || signal);
+      });
+    }
 
-  try {
-    this.child.kill(signame);
-    this.child.once('exit', function(code, signal) {
-      callback(signal || code);
-    });
-  } catch (err) {
-    if (err.code === 'ESRCH') {
-      // We got unlucky, the process is dead
-      process.nextTick(callback);
-      return;
+    for (;;) {
+      try {
+        if (!isWindows || signame === 'SIGKILL')
+          // On Unix, always use `ChildProcess#kill()` to send signals.
+          // On Windows, only the 'SIGKILL' signal can be sent to other processes.
+          this.child.kill(signame);
+        else
+          // Use the control channel to notify the child about this signal.
+          this.ctl.notify({cmd: 'signal', signame: signame, pid: this.child.pid});
+
+        break;
+
+      } catch (err) {
+        if (err.code === 'ESRCH')
+          // The process has exited in the meantime.
+          break;
+        else if (signame === 'SIGKILL')
+          // We already sent a SIGKILL so we can't get more forceful.
+          break;
```

Comment:
in these break cases, callback will never occur, it looks to me.

The contract of this method is that callback will always occur, either because it was already dead (in which case the callback value will `undefined`, that's ok), or because the exit status was reaped, in which case the callback value will be the `signo || code`.